### PR TITLE
fix(votes): always render vote-cast drawer header with close button (#2350)

### LINKED ISSUE #2350: Cast vote drawer opened from My Pending Actions is missing the vote header
## Problem

When a user opens a vote from the My Pending Actions list on the Me lens dashboard, the cast-vote drawer opens with no header at all: no vote title, no group badge, no voting status, no due date, and not even a close button. The same drawer opened from the My Votes page shows all of these details. Users arriving from Pending Actions lose the context of which vote they are casting and when it closes.

## Outcome

The cast-vote drawer shows the same header — vote title, group badge, voting status, and due date — whether it is opened from My Pending Actions or from My Votes, so users always know what they are voting on and when it closes.

## Success criteria

- Opening a vote from a My Pending Actions row shows the vote title, group, status, and due date at the top of the cast-vote drawer.
- The header matches what the My Votes page shows for the same vote.
- A close control is always visible, including while vote details are loading and if loading fails.
- Casting a vote from My Pending Actions continues to work end to end.

### PR DESCRIPTION
**Pull request by audigregorie**
## Summary

The Cast Vote drawer's header — vote title, metadata, and close button — was nested inside the `@if (vote())` block, so it only rendered once vote details had loaded. While loading, the panel opened with no header at all (no visible way to close it), and the error state had to maintain its own duplicate close-only header template. This PR hoists the header out of the vote conditional so it projects in every state, with skeleton placeholders while loading.

Resolves #2350

## Behavior changes

| Before | After |
|---|---|
| While vote details were loading, the Cast Vote panel opened with no header — no title and no close button. If loading failed, a separate stripped-down header appeared that contained only a close button. | The panel header is always visible: it shows placeholder bars while the vote loads, then the vote title and metadata once loaded. The close button is available in every state — loading, loaded, and error. |

## Technical changes

`apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html` — Vote cast drawer template

- Moved the `#header` template above the `@if (vote())` block so it always projects, regardless of vote/loading/error state
- Title slot renders the vote name when loaded, a `p-skeleton` bar while `loadingVote()`, or an empty span otherwise — the close button stays rendered beside it in all cases
- Metadata row (committee badge, vote count, due date) remains vote-only, with a skeleton bar while loading
- Removed the error branch's duplicate close-only `#header` template (including its `vote-cast-drawer-close-error` testid), since the shared header now covers that state

### CODE REVIEW

**Comment by dealako** [comment]
Hi @audigregorie 👋

Thanks for the clean fix here. This is a focused template-only change: hoisting the `#header` `ng-template` out of the `@if (vote())` block so the Cast Vote drawer always has a visible header (and close button) while loading and on error, instead of leaving the panel unclosable during load or duplicating a stripped-down error-only header. The skeleton placeholders for the title/metadata rows while `loadingVote()` are a nice touch, and removing the duplicate error-branch header (and its `vote-cast-drawer-close-error` testid) is the right cleanup now that the shared header covers that case.

I ran three parallel reviewers (security/privacy, correctness/performance/tests, style/docs) against the diff. No blocking or security/privacy issues came up — this is a low-risk UI template change with no new data flow. Cursor Bugbot's automated summary agrees (low risk, template-only UX change).

**Structured issue count:**
- 🔴 Blocking: 0 issues
- 🟡 Minor: 1 issue: missing e2e coverage for the new always-visible header across loading/loaded/error states
- ⚪ Nit: 1 issue: empty fallback `<span></span>` title placeholder in the error state has no semantic content
- ❔ Question: 0 items

**Final decision: ✅ Approved with minor comments**

Nice fix for #2350 — the behavior change (closeable in every state) is exactly what the linked issue called for.

File: apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html [hunk 1]
```diff
@@ -8,23 +8,29 @@
...
-            class="p-2 text-slate-400 hover:text-slate-600 hover:bg-slate-100 rounded-md transition-colors flex-shrink-0"
-            data-testid="vote-cast-drawer-close"
-            aria-label="Close panel">
-            <i class="fa-light fa-xmark text-xl"></i>
-          </button>
-        </div>
+        } @else if (loadingVote()) {
+          <p-skeleton width="60%" height="24px"></p-skeleton>
+        } @else {
+          <span></span>
```
Comment (dealako) [inline apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html]: [nit] Empty fallback title has no semantic content

Issue: When `vote()` is null and `!loadingVote()` (the error state), the title area renders an empty `<span></span>` with no text or aria-label.
Why it matters: The header shows only the close button and blank space in the error state, which may look like a layout glitch rather than an intentional design.
Fix: Either hide the title element entirely in this branch, or render a short placeholder (e.g. "Vote Details") so the header isn't visually empty.

File: apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html [hunk 2]
```diff
@@ -8,23 +8,29 @@
   [showCloseIcon]="false"
   styleClass="xl:w-[45%] lg:w-[55%] md:w-[70%] sm:w-[90%] w-full vote-cast-drawer"
   data-testid="vote-cast-drawer">
-  @if (vote(); as voteData) {
-    <ng-template #header>
-      <div class="flex flex-col gap-3 w-full pr-4" data-testid="vote-cast-drawer-header">
-        <div class="flex items-center justify-between gap-2">
+  <ng-template #header>
```
Comment (dealako) [inline apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html:11]: [minor] Missing e2e coverage for the always-visible header states

Issue: This PR changes the drawer to always render a header (with close button) across loading, loaded, and error states, but there's no e2e test file for this component (no `vote-cast-drawer*.spec.ts` under `apps/lfx-one/e2e/`), so the fix for #2350 (drawer previously unclosable while loading) isn't regression-protected.
Why it matters: A future change to `loadingVote()`/`vote()` signal logic could silently reintroduce the original bug (no close button during load) with nothing catching it.
Fix: Add an e2e spec covering: header + close button visible while loading (skeletons render), close button clickable during load, and close button still present in the error state. The `attendance-reconciliation-drawer.spec.ts` and `profile-edit-drawer.spec.ts` specs are reasonable patterns to follow for drawer fixture/stubbing.

**Review by dealako — DISMISSED** [review]
Approved with minor comments — see inline notes and the summary comment for the full consolidated review (security/privacy, correctness/performance/tests, style/docs all reviewed in parallel; no blocking or privacy findings).

File: apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html (hunk 2 above)
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html:11]: Agreed this deserves regression coverage. Repo convention (`.claude/rules/angular-standards.md`) is that agent-authored changes don't add Playwright specs, and stubbing the three drawer states is larger than this header fix — I'd rather track it as a follow-up issue than grow this PR. Verified manually: header + close visible while loading, close works during load, close present in the error state.

File: apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html (hunk 1 above)
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html]: Fixed in 54bc46b87 — the error-state branch now renders a 'Vote Details' placeholder h1 (same styling and testid as the loaded title) instead of an empty span, so the header keeps semantic content and the close button stays right-aligned.

**Comment by dealako** [comment]
Hi @audigregorie 👋 thanks for the quick turnaround on the feedback.

👏 **Nice work** — the empty `<span></span>` fallback in the error state is now a proper `<h1>Vote Details</h1>` using the same styling and `data-testid="vote-cast-drawer-title"` as the loaded-state heading, so the header no longer looks like a layout glitch when the vote fails to load. Clean, minimal, one-line fix that matches the existing pattern exactly.

I re-ran the three parallel reviewers (security/privacy, correctness/performance/tests, style/docs) scoped to just the new commit (`54bc46b8`). No new issues turned up, and no new bot comments have landed since the last round.

**Revision tracking:**
- ✅ **Resolved**: Empty fallback title (`<span></span>`) now renders "Vote Details" — semantic and styled consistently with the loaded title.
- ❌ **Still open**: Missing e2e test coverage for the always-visible header across loading/loaded/error states. This commit didn't add a `vote-cast-drawer` e2e spec, so the regression this PR fixes (drawer unclosable while loading, per #2350) still has no automated protection. Per our review process this carries forward as blocking since it was flagged last round and remains unaddressed.

**Structured issue count:**
- 🔴 Blocking: 1 issue: missing e2e coverage for the vote-cast-drawer header states (carried over, elevated from minor since still unaddressed)
- 🟡 Minor: 0 issues
- ⚪ Nit: 0 issues

**Final decision: 🔴 Needs changes before approval**

This is purely about test coverage — the fix itself looks correct and the prior nit is resolved. Once an e2e spec covers the loading/error header states (or you make the case that coverage lives elsewhere), this should be good to merge.

File: apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html (hunk 2 above)
Comment (dealako) [inline apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html:11]: [blocking] Still open: missing e2e coverage for the always-visible header states

Issue: Carried over from the last review round. This PR changes the drawer to always render a header (with close button) across loading, loaded, and error states — fixing #2350 (drawer previously unclosable while loading) — but there's still no e2e spec for `vote-cast-drawer` (no `vote-cast-drawer*.spec.ts` under `apps/lfx-one/e2e/`), and this commit didn't add one.
Why it matters: Without a regression test, a future change to `loadingVote()`/`vote()` signal logic could silently reintroduce the original bug with nothing catching it.
Fix: Add an e2e spec covering: header + close button visible while loading (skeletons render), close button clickable during load, and header/close button still present with the "Vote Details" fallback title in the error state. `attendance-reconciliation-drawer.spec.ts` and `profile-edit-drawer.spec.ts` are reasonable patterns to follow.

**Review by dealako — CHANGES_REQUESTED** [review]
Follow-up review: the prior nit (empty fallback title) is resolved — nice fix. The prior minor item (missing e2e coverage for the vote-cast-drawer header states) is still open and carries forward as blocking per our review process, since it was flagged last round and remains unaddressed. No new security/privacy/correctness/style issues in this commit. See the summary comment for the full breakdown.

File: apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html (hunk 2 above)
Comment (audigregorie) [reply apps/lfx-one/src/app/modules/votes/components/vote-cast-drawer/vote-cast-drawer.component.html:11]: Fixed in 719b61b26 — added `apps/lfx-one/e2e/vote-cast-drawer.spec.ts` (3 specs, green on chromium): header + close button visible while the detail request is pending (content skeletons render), close-during-load dismisses the drawer mid-flight, and a 500 detail response keeps the header/close on the list-row fallback. One honest scope note: the literal "Vote Details" error block isn't drivable through a real UI flow — the votes dashboard always passes `listVote` (`catchError` falls back to it) and the dashboard-cast-drawer-host pre-fetches through the same 10s shareReplay cache (a prefetch failure shows a toast and never opens the drawer) — so the spec asserts the reachable error behavior and documents this in the file header.

File: apps/lfx-one/e2e/vote-cast-drawer.spec.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+test.describe('Vote cast drawer — always-visible header (GH-2350)', () => {
+  test('keeps the header and close button visible while the detail loads, then renders the ballot', async ({ page }) => {
+    const gate = deferred<null>();
+    await page.route(`**/api/votes/${VOTE_UID}`, async (route) => {
+      await gate.promise;
+      return fulfillJson(route, buildVoteDetail());
+    });
+    await gotoMyVotes(page);
+
+    await page.getByTestId(`votes-cast-${VOTE_UID}`).click();
```
[the remark this answers is not in the document]
Comment (audigregorie) [reply apps/lfx-one/e2e/vote-cast-drawer.spec.ts:170]: Fixed in 0d71631 — added `vote-cast-drawer.component.spec.ts` (TestBed) mounting the drawer with `listVote = null`: the header + close button render while a deferred `getVote()` is pending (then the ballot swaps in), the drawer closes mid-load, and a failed `getVote()` shows the "Vote Details" fallback title + error block. Verified all three tests fail against the pre-fix template and pass on the fixed one — true GH-2350 regression coverage.

**Comment by dealako** [comment]
Hi @audigregorie 👋 great follow-through here.

👏 **Nice work** — `apps/lfx-one/e2e/vote-cast-drawer.spec.ts` and the new `vote-cast-drawer.component.spec.ts` unit spec together give solid, substantive coverage of exactly the regression this PR fixes: header + close button visible and clickable while loading, close button working mid-request, and the error-state fallback ("Vote Details" title) staying present alongside a working close button. The unit spec nicely fills in the null-`listVote` edge case that the e2e flow can't reach on its own, and both files carry proper license headers, doc comments explaining *why* the two layers complement each other, and match this repo's existing drawer-spec conventions (e.g. `attendance-reconciliation-drawer.spec.ts`).

I re-ran the three parallel reviewers scoped to the two new commits (`719b61b26`, `0d71631a1`). No new issues surfaced, and the one carried-over blocking item is now resolved.

**Revision tracking:**
- ✅ **Resolved**: Missing e2e coverage for the vote-cast-drawer header states — now covered end-to-end (loading, loaded, error) plus a unit-test layer for the null-`listVote` edge case.

**Structured issue count:**
- 🔴 Blocking: 0 issues
- 🟡 Minor: 0 issues
- ⚪ Nit: 0 issues

**Final decision: ✅ Approved**

Thanks for closing this out — good fix for #2350 with the test coverage to back it up.

**Review by dealako — APPROVED** [review]
Follow-up review: the remaining blocking item (missing e2e coverage for the vote-cast-drawer header states) is now resolved by the two new test files. No new issues found in the security/privacy, correctness/performance/tests, or style/docs passes. See the summary comment for the full breakdown.
